design/include-cleaner.md

```diff
@@ -0,0 +1,167 @@
+# Include Cleaner
+
+Manually managing includes in a C++ translation unit, especially in the face of
+transitive inclusions, requires a lot of effort. Include Cleaner aims to
+provide diagnostics to keep includes in an
+[IWYU](https://include-what-you-use.org/)-clean state.
+
+Include Cleaner is a IWYU-inspired feature in clangd that automates the process
+of maintaining a necessary set of included headers. It now is available in
+"preview" mode with an incomplete set of capabilities and can be enabled
+through [configuration file](/config#UnusedIncludes). If you experience any
+bugs, please submit a bug report in
+[clangd/issues](https://github.com/clangd/clangd/issues).
+
+{:.v14}
+
+## How it works
+
+Include Cleaner issues diagnostics fro includes that are present but not used
+in the main file. When you open a file, clangd will analyze the symbols the
+file uses and mark all headers defining these symbools as "used". The warnings
+will be issues for the headers that are included but not marked as "used".
+Example:
+
+```c++
+// foo.h
+class Foo {};
+```
+
+```c++
+// bar.h
+class Bar {};
+```
+
+```c++
+// main.cpp
+
+#include "foo.h"
+#include "bar.h" // <- Will be marked as unused and suggested to be removed.
+
+int main() {
+  Foo f;
+}
+```
+
+Here, `main.cpp` only makes use of symbols from `foo.h` and removing `#include
+"bar.h"` prevents unnecessary parsing of `bar.h` and allows breaking the
+dependncy on it.
+
+### Deciding what headers are used
+
+Clangd relies on Clang AST to deduce which headers are used and which aren't,
+the whole Include Cleaner decision process is described below.
+
+#### Scanning the main file
+
+First, Include Cleaner will build the AST for the main file (the file currently
+opened in the editor). After the AST is built, Include Cleaner will recursively
+visit the AST nodes and collect locations of all referenced symbols (e.g.
+types, functions, global variables). The AST will contain macro and `auto`
+expansions. For example, in this case, `foo.h` will be considered as used:
+
+```c++
+// foo.h
+
+int foo() { return 42; }
+```
+
+```c++
+// macro.h
+
+#include "foo.h"
+
+#define FOO foo
+```
+
+```c++
+// main.cpp
+
+#include "macro.h"
+
+int main() {
+  FOO();
+}
+```
+
+This means that Include Cleaner is conservatively treating symbols in the
+expanded code as usages as opposed to only explicitly spelled symbols.
+
+Include Cleaner will also traverse the unexpanded macro tokens to see where
```

Comment:
i am not sure what it means to be `unexpanded`. are we trying to say macro names mentioned in the spelled code?